### PR TITLE
Add a region caches

### PR DESCRIPTION
**Pull request by nrc**
Builds on https://github.com/nrc/client-rust/pull/new/region-cache. It is pretty hard to test this. I think we need integration tests on mock TiKV so that we can cause region failures. There are some unit tests, but not as many as I would like.

PTAL @sticnarf @Hoverbear

### CODE REVIEW

File: src/mock.rs [hunk 1]
```diff
@@ -111,7 +111,7 @@ impl PdClient for MockPdClient {
         ))))
     }
 
-    fn region_for_key(&self, key: &Key) -> BoxFuture<'static, Result<Region>> {
+    fn region_for_key(&self, key: &Key, _: bool) -> BoxFuture<'static, Result<Region>> {
```
Comment (Hoverbear) [inline src/mock.rs]:
```suggestion
    fn region_for_key(&self, key: &Key, _skip_cache: bool) -> BoxFuture<'static, Result<Region>> {
```

File: src/mock.rs [hunk 2]
```diff
@@ -122,7 +122,7 @@ impl PdClient for MockPdClient {
         Box::pin(ready(Ok(region)))
     }
 
-    fn region_for_id(&self, id: RegionId) -> BoxFuture<'static, Result<Region>> {
+    fn region_for_id(&self, id: RegionId, _: bool) -> BoxFuture<'static, Result<Region>> {
```
Comment (Hoverbear) [inline src/mock.rs]:
```suggestion
    fn region_for_id(&self, id: RegionId, _skip_cache: bool) -> BoxFuture<'static, Result<Region>> {
```

File: src/request.rs [hunk 3]
```diff
@@ -44,7 +44,7 @@ pub trait KvRequest: Sync + Send + 'static + Sized {
         mut self,
         pd_client: Arc<impl PdClient>,
     ) -> BoxStream<'static, Result<Self::RpcResponse>> {
-        let stores = self.store_stream(pd_client.clone());
+        let stores = self.store_stream(pd_client.clone(), false);
```
Comment (sticnarf) [inline src/request.rs]: I think it's not correct to always set it to false. Parameter `is_retry` should be also added to this method and set `is_retry` to `true` on region error.

**Comment by nrc** [comment]
Comments addressed, PTAL @sticnarf @Hoverbear

File: src/pd/retry.rs [hunk 4]
```diff
@@ -66,21 +69,47 @@ impl RetryClient<Cluster> {
...
             timeout,
+            cache: RwLock::new(RegionCache::new()),
         })
     }
 
     // These get_* functions will try multiple times to make a request, reconnecting as necessary.
-    pub async fn get_region(self: Arc<Self>, key: Vec<u8>) -> Result<Region> {
+    pub async fn get_region(self: Arc<Self>, key: Key, skip_cache: bool) -> Result<Region> {
+        if skip_cache {
+            self.cache.write().unwrap().remove_region_by_start_key(&key);
```
Comment (sticnarf) [inline src/pd/retry.rs]: The `key` here isn't always the start key of a region. Maybe we can add a new method to the region cache.

**Comment by Hoverbear** [comment]
So this fixes #114 ?

**Comment by sticnarf** [comment]
> So this fixes #114 ?

If this PR gets merged (without changing a lot), maybe #114 can be closed but we need to create new issues to track the unsolved problems.

File: src/pd/retry.rs (hunk 4 above)
Comment (Hoverbear) [reply src/pd/retry.rs]: @nrc PTAL

**Comment by sadroeck** [comment]
Any update on this?

**Comment by nrc** [comment]
I'll come back to this some day, but closing for now since I don't have time.

**Pull request closed without merging**